policy/reviewer/reviewer.go

```diff
@@ -0,0 +1,162 @@
+			if j > n*5 {
+				// We haven't been able to select a random value, bail loudly
+				panic(fmt.Sprintf("Unable to select random value for %d %d", n, len(users)))
+			}
+			m := r.Intn(len(users))
+			if !selected[m] {
+				selected[m] = true
+				selections = append(selections, users[m])
+				break
+			}
+			j++
+		}
+	}
+	return selections
+}
+
+func shoveIntoMap(m map[string]struct{}, u []string) {
+	for _, n := range u {
+		m[n] = struct{}{}
+	}
+}
+
+func selectTeamMembers(prctx pull.Context, allTeams []string, r *rand.Rand) ([]string, error) {
+	randomTeam := allTeams[r.Intn(len(allTeams))]
+	teamInfo := strings.Split(randomTeam, "/")
+	teamMembers, err := prctx.ListTeamMembers(teamInfo[0], teamInfo[1])
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to get member listing for team %s", randomTeam)
+	}
+	return teamMembers, nil
+}
+
+func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common.Result, r *rand.Rand) ([]string, error) {
+	logger := zerolog.Ctx(ctx)
+	pendingLeafNodes := findLeafChildren(result)
+	var requestedUsers []string
+
+	logger.Debug().Msgf("Collecting reviewers for %d pending leaf nodes", len(pendingLeafNodes))
+
+	for _, child := range pendingLeafNodes {
+		allUsers := make(map[string]struct{})
+		shoveIntoMap(allUsers, child.ReviewRequestRule.Users)
+
+		if len(child.ReviewRequestRule.Teams) > 0 {
+			teamMembers, err := selectTeamMembers(prctx, child.ReviewRequestRule.Teams, r)
+			if err != nil {
+				logger.Warn().Err(err).Msgf("Unable to get member listing for teams, skipping team member selection")
+			}
+			shoveIntoMap(allUsers, teamMembers)
+		}
+
+		if len(child.ReviewRequestRule.Organizations) > 0 {
+			randomOrg := child.ReviewRequestRule.Organizations[r.Intn(len(child.ReviewRequestRule.Organizations))]
+			orgMembers, err := prctx.ListOrganizationMembers(randomOrg)
+			if err != nil {
+				logger.Warn().Err(err).Msgf("Unable to get member listing for org %s, skipping org member selection", randomOrg)
+			}
+			shoveIntoMap(allUsers, orgMembers)
+		}
+
+		allCollaboratorPermissions, err := prctx.ListRepositoryCollaborators()
+		if err != nil {
+			return nil, errors.Wrap(err, "Unable to list repository collaborators")
+		}
+
+		if child.ReviewRequestRule.Admins {
+			var repoAdmins []string
+			for user, perm := range allCollaboratorPermissions {
+				if perm == common.GithubAdminPermission {
+					repoAdmins = append(repoAdmins, user)
+				}
+			}
+			shoveIntoMap(allUsers, repoAdmins)
+		}
+
+		if child.ReviewRequestRule.WriteCollaborators {
+			var repoCollaborators []string
+			for user, perm := range allCollaboratorPermissions {
+				if perm == common.GithubWritePermission {
+					repoCollaborators = append(repoCollaborators, user)
+				}
+			}
+			shoveIntoMap(allUsers, repoCollaborators)
+		}
+
+		var allUserList []string
+		for u := range allUsers {
+			// Remove any users who aren't collaborators or the author, since github will fail to assign _anyone_
```

Comment:
style: `Remove the author and any users who aren't collaborators...`; it was unclear to me if "aren't" applied to "being the author" in the original order.